src/router_node.c

```diff
@@ -1829,11 +1838,16 @@ static void CORE_delivery_update(void *context, qdr_delivery_t *dlv, uint64_t di
         // If the delivery is still arriving, don't push out the disposition change yet.
         //
         if (qd_message_receive_complete(msg)) {
-            pn_delivery_update(pnd, disp);
+            if (disp != pn_delivery_local_state(pnd)) {
+                pn_delivery_update(pnd, disp);
+                if (qdr_delivery_disposition(dlv) != disp)
```

Comment:
This should not be necessary (and it's a race): The disp passed into this function comes from the qdr_delivery_t so this check and set is a no-op.  I would remove these.